Simplify VisitArrow

ddn0 opened the pull request:
I added `CouldAccept` to support multiple arguments better and it works well enough on a few number of arguments assuming that most instantiations can be pruned.

I'm not super happy with how parameter types ("caller" types) are added to VisitArrow
https://github.com/KatanaGraph/katana/pull/536/files#diff-530f8996865d146ccf2fe0d9269f3e551b7cf2883fb01d75e50474a9011505d3R60 in this PR.

The current way doesn't handle cases where the caller may have a mix of downcasted and general types, e.g., `VisitArrow(visitor, arrow::Scalar, arrow::StringScalar)`. I'm thinking of reworking VisitArrow to call some function on each one of its arguments to allow callers to customize the casting rules on input. This should also remove the need for all these decays everywhere as this is normalize the caller types as well.

Note the previous implementation required homogeneous parameter types, which seems like an odd restriction once we support multiple parameters. 

Anyway, the state is good enough to get some feedback and I'm open to thoughts on how to improve the interface since I have been mucking around with VisitArrow anyway.

```
    libsupport: Simplify VisitArrow

    Switch to simplified VisitArrow implementation, which supports template
    instantiation pruning, ensures no intermediate or shared_ptr copying and
    reduces the number of template variables (for improved readability).

    Performance results below. The current implementation is now called
    Visit. See previous commit for performance of previous implementation.

      spork$ external/katana/libsupport/test/arrow-bench  --benchmark_filter=/1048576
      2021-09-10T01:14:23-07:00
      Running external/katana/libsupport/test/arrow-bench
      Run on (12 X 4500 MHz CPU s)
      CPU Caches:
        L1 Data 32 KiB (x6)
        L1 Instruction 32 KiB (x6)
        L2 Unified 256 KiB (x6)
        L3 Unified 12288 KiB (x1)
      Load Average: 0.27, 0.56, 1.00
      ***WARNING*** CPU scaling is enabled, the benchmark real time measurements may be noisy and will incur extra overhead.
      -----------------------------------------------------------------------------------
      Benchmark                         Time             CPU   Iterations UserCounters...
      -----------------------------------------------------------------------------------
      InlineSwitch/1048576        4621561 ns      4621457 ns          156 items_per_second=226.893M/s
      Visit/1048576              13067725 ns     13067640 ns           53 items_per_second=80.2422M/s
      DynamicCast/1048576       301830300 ns    301824285 ns            2 items_per_second=3.47413M/s
      SwitchCast/1048576         11785539 ns     11785325 ns           64 items_per_second=88.973M/s
      SwitchCastResult/1048576   11102731 ns     11102656 ns           63 items_per_second=94.4437M/s
```

```
    libsupport: Add VisitArrow benchmark

    In preparation of refactoring katana::VisitArrow, benchmark various
    alternative implementation strategies.

    The variants compared are:

    - Visit is the current implementation
    - InlineSwitch is a hand-optimized version, a performance ceiling.
    - DynamicCast, SwitchCast, SwitchCastResult are possible implementation
      strategies
    - Dispatch is the proposed alternative to Visit

    Below are the performance results.

      spork$ external/katana/libsupport/test/arrow-bench  --benchmark_filter=/1048576
      2021-09-10T01:22:47-07:00
      Running external/katana/libsupport/test/arrow-bench
      Run on (12 X 4500 MHz CPU s)
      CPU Caches:
        L1 Data 32 KiB (x6)
        L1 Instruction 32 KiB (x6)
        L2 Unified 256 KiB (x6)
        L3 Unified 12288 KiB (x1)
      Load Average: 3.41, 1.50, 1.12
      ***WARNING*** CPU scaling is enabled, the benchmark real time measurements may be noisy and will incur extra overhead.
      -----------------------------------------------------------------------------------
      Benchmark                         Time             CPU   Iterations UserCounters...
      -----------------------------------------------------------------------------------
      InlineSwitch/1048576        4315897 ns      4315869 ns          163 items_per_second=242.958M/s
      Visit/1048576              14571726 ns     14571503 ns           48 items_per_second=71.9607M/s
      DynamicCast/1048576       290699798 ns    290694342 ns            2 items_per_second=3.60714M/s
      SwitchCast/1048576         10870668 ns     10870484 ns           64 items_per_second=96.4608M/s
      SwitchCastResult/1048576   10878910 ns     10878452 ns           65 items_per_second=96.3902M/s
      Dispatch/1048576           11092319 ns     11092303 ns           63 items_per_second=94.5319M/s

    After some investigation, the main reason for the difference between
    Visit and the other switch-based implementations (include Dispatch) is
    due to excessive copying of std::shared_ptr<arrow::DataType> in
    Type(const arrow::Scalar&). Once this is fixed, all variants have
    roughly the same throughput.

    The Dispatch implementation is still an improvement over Visit in terms
    of readability and maintainability and it supports multi-arguments
    better due to its support for pruning instantiations.

```

roshandathathri commented on libdist/include/galois/runtime/Network.h:
```diff
@@ -78,27 +77,21 @@ class GALOIS_EXPORT NetworkInterface : public CommBackend {
+  //! landing pad and some data (buf).
+  //! On the receiver, recv(buf) will be called during handleRecvs()
   //! buf is invalidated by this operation
   void sendMsg(
```
I would prefer this function to be dropped. We haven't used it in a long time. This was designed for the _active message_ model of communication and I don't see a use for it in the immediate future.

roshandathathri commented on libdist/include/galois/runtime/Network.h:
```diff
@@ -78,27 +77,21 @@ class GALOIS_EXPORT NetworkInterface : public CommBackend {
+  //! on the receiver, buf will be returned on a recvTagged(tag)
   //! buf is invalidated by this operation
   virtual void sendTagged(
       uint32_t dest, uint32_t tag, SendBuffer& buf, int type = 0) = 0;
```
`SendBuffer&& buf` is precise. We can defer that as that will involve changing a lot of code.

ddn0 replied:
Turns out it wasn't too much code, so I added this change as another commit at the end of this one.

roshandathathri replied:
Awesome!

tylershunt reviewed (approved):
LGTM I always spell Receive wrong then fix it only to realize I had it right the first time.

roshandathathri commented on libdist/include/galois/runtime/Network.h:
```diff
@@ -41,10 +41,9 @@
-using RecvBuffer = DeSerializeBuffer;
+//! The result of receiving from a network interface is a pair of sending host
+//! and data.
+using RecvResult = std::optional<std::pair<uint32_t, RecvBuffer>>;
```
Isn't this better as a type within the NetworkInterface class?

ddn0 replied:
It's a little hard to refer to if it is in the NetworkInterface class:

```
galois::runtime::NetworkInterface::RecvResult
// or should I use the concrete subclass?
galois::runtime::NetworkInterfaceBuffered::RecvResult
```
vs
```
galois::runtime::RecvResult
```
which makes it clear that this is independent of the implementation class.

This is not big deal to me, but that was my reasoning.

roshandathathri replied:
Ok, makes sense. SendBuffer and RecvBuffer are in galois::runtime as well.

roshandathathri commented on libdist/include/galois/runtime/Serialize.h:
```diff
@@ -52,57 +52,30 @@ namespace runtime {
   //! the actual data stored in this buffer
   vTy bufdata;
-  int offset;
+  unsigned offset{0};
```
Why not uint32_t?

ddn0 replied:
cool

roshandathathri replied:
Cool

roshandathathri commented on libdist/include/galois/runtime/Serialize.h:
```diff
@@ -174,27 +148,19 @@ class DeSerializeBuffer {
 
 public:
-  //! default constructor
-  SerializeBuffer() = default;
```
I'm surprised that this isn't needed?

roshandathathri replied:
Thanks for the clarification.

roshandathathri commented on libdist/include/galois/runtime/Serialize.h:
```diff
@@ -174,27 +148,19 @@ class DeSerializeBuffer {
+  //! Moves DeserializedBuffer into SendBuffer
+  SendBuffer& operator=(RecvBuffer&& rhs) {
+    auto buf = std::move(rhs);
+    bufdata = std::move(buf.getVec());
```
What's the difference b/w this and `rhs.getVec()`?

roshandathathri replied:
Still confusing to me. `rhs` might be destructed earlier but then `buf` will only be destructed when the function returns, right?

ddn0 replied:
```cpp
// Version 1
SendBuffer& operator=(RecvBuffer&& a) {
  bufdata = std::move(a.getVec());
  ...
}

// Version 2
SendBuffer& operator=(RecvBuffer&& a) {
  auto buf = std::move(a);
  bufdata = std::move(buf.getVec());
  ...
}

void Fn() {
  RecvBuffer rbuf;
  SendBuffer b = std::move(rbuf);
  .....
}
```

In version 1, if RecvBuffer had any other fields besides `getVec` they would not be released until `Fn` returns. In version 2, those fields would be released when `operator=(RecvBuffer&&)` returns.

roshandathathri replied:
Thanks for the clarification. Didn't know this. I thought that the move object passed as a function parameter would be destructed at the end of that function, which in this case is `operator=`.

ddn0 replied:
I can't claim to be an expert in move semantics, but the model that helps me is that `std::move` is just a cast to an rvalue, and an rvalue is just a type that helps dispatch to the operations that actually implement move behavior. It isn't until you use `operator=(T&&)` or similar that the magic happens. In fact, it is probably a bug to take an rvalue as a parameter and never `std::move` it or use a sink function like `operator=(T&&)`.

roshandathathri replied:
Thanks!

roshandathathri reviewed (approved):
This is great!

Some minor nits below.

gurbinder533 reviewed (approved):
LGTM.

ddn0 committed:
libsupport: Add VisitArrow benchmark

In preparation of refactoring katana::VisitArrow, benchmark various
alternative implementation strategies.

The variants compared are:

- Visit is the current implementation
- InlineSwitch is a hand-optimized version, a performance ceiling.
- DynamicCast, SwitchCast, SwitchCastResult are possible implementation
  strategies
- Dispatch is the proposed alternative to Visit

Below are the performance results.

  spork$ external/katana/libsupport/test/arrow-bench  --benchmark_filter=/1048576
  2021-09-14T14:56:57-07:00
  Running external/katana/libsupport/test/arrow-bench
  Run on (12 X 4500 MHz CPU s)
  CPU Caches:
    L1 Data 32 KiB (x6)
    L1 Instruction 32 KiB (x6)
    L2 Unified 256 KiB (x6)
    L3 Unified 12288 KiB (x1)
  Load Average: 0.77, 3.09, 4.04
  ***WARNING*** CPU scaling is enabled, the benchmark real time measurements may be noisy and will incur extra overhead.
  -----------------------------------------------------------------------------------
  Benchmark                         Time             CPU   Iterations UserCounters...
  -----------------------------------------------------------------------------------
  InlineSwitch/1048576        4439905 ns      4414797 ns          159 items_per_second=237.514M/s
  Visit/1048576              13542043 ns     13469615 ns           52 items_per_second=77.8475M/s
  DynamicCast/1048576       283761809 ns    282193524 ns            2 items_per_second=3.7158M/s
  SwitchCast/1048576         10113602 ns     10058763 ns           69 items_per_second=104.245M/s
  SwitchCastResult/1048576   10332473 ns     10276416 ns           69 items_per_second=102.037M/s
  Dispatch/1048576           11060736 ns     10999973 ns           64 items_per_second=95.3253M/s

After some investigation, the main reason for the difference between
Visit and the other switch-based implementations (include Dispatch) is
due to excessive copying of std::shared_ptr<arrow::DataType> in
Type(const arrow::Scalar&). Once this is fixed, all variants have
roughly the same throughput.

The Dispatch implementation is still an improvement over Visit in terms
of readability and maintainability and it supports multi-arguments
better due to its support for pruning instantiations.

ddn0 committed:
libsupport: Simplify VisitArrow

Switch to simplified VisitArrow implementation, which supports template
instantiation pruning, ensures no intermediate or shared_ptr copying and
reduces the number of template variables (for improved readability).

Performance results below. The current implementation is now called
Visit. See previous commit for performance of previous implementation.

  spork$ $ external/katana/libsupport/test/arrow-bench  --benchmark_filter=/1048576
  2021-09-14T14:59:49-07:00
  Running external/katana/libsupport/test/arrow-bench
  Run on (12 X 4500 MHz CPU s)
  CPU Caches:
    L1 Data 32 KiB (x6)
    L1 Instruction 32 KiB (x6)
    L2 Unified 256 KiB (x6)
    L3 Unified 12288 KiB (x1)
  Load Average: 0.79, 2.01, 3.46
  ***WARNING*** CPU scaling is enabled, the benchmark real time measurements may be noisy and will incur extra overhead.
  -----------------------------------------------------------------------------------
  Benchmark                         Time             CPU   Iterations UserCounters...
  -----------------------------------------------------------------------------------
  InlineSwitch/1048576        4187416 ns      4187355 ns          167 items_per_second=250.415M/s
  Visit/1048576              10397289 ns     10397037 ns           67 items_per_second=100.853M/s
  DynamicCast/1048576       269501814 ns    269501552 ns            3 items_per_second=3.8908M/s
  SwitchCast/1048576         10186293 ns     10185547 ns           69 items_per_second=102.947M/s
  SwitchCastResult/1048576   10103287 ns     10102396 ns           69 items_per_second=103.795M/s

danielmawhirter commented on libsupport/include/katana/ArrowVisitor.h:
```diff
@@ -35,129 +22,166 @@ using is_list_type_patched = std::integral_constant<
 template <typename T, typename R = void>
 using enable_if_list_type_patched =
-    arrow::enable_if_t<is_list_type_patched<T>::value, R>;
+    std::enable_if_t<is_list_type_patched<T>::value, R>;
```
This can be removed with https://github.com/KatanaGraph/katana-enterprise/pull/1931

danielmawhirter commented on libsupport/include/katana/ArrowVisitor.h:
```diff
@@ -35,129 +22,166 @@ using is_list_type_patched = std::integral_constant<
+    } else if constexpr (!std::decay_t<Visitor>::template CouldAccept<
+                             ArrowTypes...>()) {
+      return KATANA_ERROR(
+          katana::ErrorCode::ArrowError, "could not accept type id");
```
I see how this avoids instantiating the type-specific visitor methods. What surprises me is that this is sufficient to keep the compiler from barfing. Wouldn't this function itself still be specialized for every type-specific case?
I guess that assumes that CouldAccept returns `true` until it 'sees' every argument, but it could short-circuit to prune the space earlier?
Is this intuition remotely on the right track?

danielmawhirter replied:
A curiosity more than a request, how hard would it be to derive the behavior of CouldAccept from types accepted by the methods of the visitor itself?

rob-isaac replied:
This may be impossible (or not worth the time investment) but it would be awesome to have the visitor support specific types with `enable_if` and automatically prune if the `Call` function doesn't exist for the template specialization. Can def go in a future PR though.

ddn0 replied:
> I guess that assumes that CouldAccept returns true until it 'sees' every argument, but it could short-circuit to prune the space earlier?
> Is this intuition remotely on the right track?

Exactly. That's what I was trying to convey with the name `CouldAccept`. But yes, `CouldAccept` needs to be a  incremental and overapproximates `visitor.Call` and we get the main effect if it short-circuits.

> how hard would it be to derive the behavior of CouldAccept from types accepted by the methods of the visitor itself?

It is not possible to know if a function exists without first providing arguments due to function overloading, so I don't think it is possible to derive the information we need solely from the definition of `visitor.Call`.

This does remind me: I also considered providing a catch all case in `VisitArrow` itself, so that visitors can avoid making themselves total functions like we do now:

```cpp
struct Visitor {
  template<typename ....>
  std::enable_if_t<P, ResultType> Call(....) {}

  template<typename ....>
  std::enable_if_t<!P, ResultType> Call(....) {
    // handle error case
  }
};
```

but I thought it might be better for visitors to explicitly handle these cases to force developers to think about these cases and provide better error messages. But this gets somewhat ugly as we increase the number parameters.

Thoughts?

danielmawhirter commented on libsupport/include/katana/ArrowVisitor.h:
```diff
@@ -345,7 +338,7 @@ class AppendScalarToBuilder {
             katana::ErrorCode::ArrowError, "failed to get scalar");
       }
-      if (auto res = VisitArrow(scalar_res.ValueOrDie(), visitor); !res) {
+      if (auto res = VisitArrow(visitor, *scalar_res.ValueOrDie()); !res) {
```
`KATANA_CHECKED` for res, scalar_res

danielmawhirter commented on libsupport/include/katana/ArrowVisitor.h:
```diff
@@ -368,7 +361,7 @@ class AppendScalarToBuilder {
     for (int f = 0, n = scalar.value.size(); f < n; ++f) {
       AppendScalarToBuilder visitor(builder->field_builder(f));
-      if (auto res = VisitArrow(scalar.value.at(f), visitor); !res) {
+      if (auto res = VisitArrow(visitor, *scalar.value.at(f)); !res) {
```
`KATANA_CHECKED`

danielmawhirter reviewed (approved):
And here I was thinking I'd gotten pretty comfortable with templates. I would mention a minor preference for 'Specialized' over "Downcasted', but if downcasted is more standard then stick with it.
This is great, I love seeing this code all grown up, thanks for working on this Donald!

rob-isaac commented on libsupport/include/katana/ArrowVisitor.h:
```diff
@@ -35,129 +22,166 @@ using is_list_type_patched = std::integral_constant<
-/// 3) Call the templated visitor with the specialized parameter
+namespace internal {
+
+// decltype(auto) is the complement of perfect forwarding,
```
Very cool, did not know about this

rob-isaac commented on libsupport/include/katana/ArrowVisitor.h:
```diff
@@ -35,129 +22,166 @@ using is_list_type_patched = std::integral_constant<
+  /// with more than one call argument as the number of template instantiations
+  /// grows exponentially with the number of arguments.
+  template <typename... ArrowTypes>
+  static constexpr bool CouldAccept() {
```
May want to actually enforce implementing this function, especially for functions that take 3+ args

rob-isaac commented on libsupport/test/arrow-bench.cpp:
```diff
@@ -0,0 +1,280 @@
+#include <memory>
```
Benchmarking is good

rob-isaac reviewed (approved):
Overall much cleaner, and the testing is good. Further iterations can be added to address your comment in the PR description and potential for automatic pruning of types.

ddn0 commented:
Thinking about it more, it is probably more sensible to just create a set of categories and just make visitors define themselves in terms of these.

```cpp
using numeric_like = std::tuple<arrow::Int64Type, arrow:UInt64Type, ....>
using string_like = std::tuple<arrow::StringType, arrow::LargeStringType>
....

struct Visitor {
  using ResultType = ....;
  using Accept = std::tuple<numeric_like, ....>;

  template<typename ArrowTypes..., typename... Args>
  ResultType Call(Args&&... args) { .... }

  template<typename ArrowTypes..., typename... Args>
  ResultType AcceptFailed(Args&&... args) { .... }
};
```

No need for the funny `CouldAccept`, expansion is "linear" in the number of arguments and pulling out failures explicitly lets callers completely define it and we can use non-`katana::Result` `ResultType` s if we want.

rob-isaac commented:
> Thinking about it more, it is probably more sensible to just create a set of categories and just make visitors define themselves in terms of these.
> 
> ```c++
> using numeric_like = std::tuple<arrow::Int64Type, arrow:UInt64Type, ....>
> using string_like = std::tuple<arrow::StringType, arrow::LargeStringType>
> ....
> 
> struct Visitor {
>   using ResultType = ....;
>   using Accept = std::tuple<numeric_like, ....>;
> 
>   template<typename ArrowTypes..., typename... Args>
>   ResultType Call(Args&&... args) { .... }
> 
>   template<typename ArrowTypes..., typename... Args>
>   ResultType AcceptFailed(Args&&... args) { .... }
> };
> ```
> 
> No need for the funny `CouldAccept`, expansion is "linear" in the number of arguments and pulling out failures explicitly lets callers completely define it and we can use non-`katana::Result` `ResultType` s if we want.

I like this format. Ideally, there will be a default error thrown like `KATANA_ERROR(<error_code>,"visitor {} does not accept type(s) {}", typeid(*this), std::string(typeid(args).name() + ", " +...)`. Notes: Probably not the correct syntax for the fold expression and this will throw away "const" and "&" from type, but there are ways to print the type with those qualifiers if desired.

Also many `numeric_like`'s will still be exponential so maybe we should consider dropping unsupported numeric types like UInt16 and Int16.

ddn0 commented:
@rob-isaac , @danielmawhirter Can I get a rereview? I changed the visitor interfaces to be simpler.

You can ignore the last two commits in this PR. I am submitting PRs for those separately, and I'll remove them from here once they get merged.

danielmawhirter reviewed (approved):
I'm happy, this looks like a cleaner interface overall.

rob-isaac commented:
> I'm happy, this looks like a cleaner interface overall.

Agreed
